addons/website/static/src/snippets/s_table_of_content/options.js

```diff
@@ -18,6 +18,7 @@ options.registry.TableOfContent = options.Class.extend({
         const config = {attributes: false, childList: true, subtree: true, characterData: true};
         this.observer = new MutationObserver(() => this._generateNav());
         this.observer.observe(targetNode, config);
+        this.$target.on('content_changed', () => this._generateNav());
```

Comment:
> This is only ocuring when you have a single entry in the table of content?

Hello @rdeodoo ,
No, it will also reproduce for multiple entries as we can see [here](https://49712776-15-0-all.runbot156.odoo.com/#scrollTop=0). I have reproduced it by removing `overflow-hidden`. Can you please check it out?

Reply:
> // The mutation observer doesn't observe the attributes change, it would
>         // be too much. Adding content_changed "listener" instead.

Done the changes . Thanks

Reply:
Ok I see what's going on here.
The `0px` css rule overriding the `-1px` rule for the last item is not being applied anymore if the last element is invisible.
![image](https://github.com/odoo/odoo/assets/30048408/81b70e5f-30cd-4c90-9a92-5a5d030f5f18)

The class you are adding will just fix the consequence of this behavior and not the root cause.
You need to find a way to make it so the 0px is applied to the last **visible** one.
I don't think you can make it in css directly as there is no selector for that, neither indirectly in css as there is no d-none class to do something like `:not(.d-none):last-child`.

There is surely no solution in css, as to know which elements to hide regarding the visibility stuff, it simply do things like:
```scss
html:not([data-utm-campaign="Sale"]) body:not(.editor_enable) [data-visibility-id="utm-campaign_o_1"] {
    display: none !important;
}
```

Maybe the TOC system could have some special JS code in its `000.js` `start()` method to add a class to the last element of the TOC like `o_table_of_content_link_last` which would come with a scss rule to add the `margin-bottom: 0` to the real visible last element.
Without forgetting to remove this class in the `destroy`.

Reply:
@rdeodoo To answer here and to what you said in the main review comment: I think you summed up nearly everything. In stable, I think we have no other solution than to find the first and last visible elements, mark them with a custom class and re-create bootstrap CSS rules to target those.

In master (where this specific bug may not appear anymore with BS5 but where there are also rules that target the first and last elements (for the border-radius)), I guess a better solution would be to **remove** the invisible elements in the list-group area at `start` and attach them back in `destroy`. That way, we don't have to duplicate bootstrap CSS rules. Arguably, we could do that directly in stable if you think it is ok @rdeodoo ... it is just more "risky" ("what if there is custom JS code which relied on those invisible elements to be in the DOM for whatever reason"). Maybe a less risky way for stable would be to keep them in the list-group but switch them to a different position (like before the last visible one) and put them back in the right order in `destroy`, but maybe I am overthinking things.

Reply:
Ty :+1: 

So, long story short: add class to visible last and target it through scss for less risk, or move the visible last as real last element so no scss or new class is needed but a bit riskier.

I think it's ok to do that in stable IMHO, I don't think it's overthinking, it seems to be a good compromise :+1: 
About removing or reordering, I would let @utag-odoo pick what seems the easiest to develop.
In both cases, he would need to put them back where they should be on destroy, so keeping the order somewhere for both cases.

Reply:
> I guess a better solution would be to **remove** the invisible elements in the list-group area at `start` and attach them back in `destroy`.

Hello @qsm-odoo ,
I believe this solution may not be effective because attaching the invisible elements to the list-group area initially and then removing them during the destroy process does not seem to take into account the behavior associated with the `"data-visibility-id"`. When we switch the language to the visibility language, these elements also become hidden at that point.

Reply:
Let's continue the discussion at https://github.com/odoo/odoo/pull/126705#discussion_r1305691593 :shrug: